text/0000-dyn-type.md

```diff
@@ -0,0 +1,2261 @@
+```
+
+Modify the `core::marker::Sized` marker trait to:
+
+```rust ,ignore
+#[stable(feature = "rust1", since = "1.0.0")]
+#[lang = "sized"]
+#[rustc_on_unimplemented = "`{Self}` does not have a constant size known at compile-time"]
+#[fundamental]
+pub trait Sized: Aligned {}
+//               ^~~~~~~ new
+```
+
+Trying to implement `Pointee`, `Derived`, `Aligned` or `Sized` should emit error [E0322]
+*“explicit impls for the trait are not permitted”*. Only `Regular` and `Inline` are designed to be customized.
+
+[E0322]: https://doc.rust-lang.org/error-index.html#E0322
+
+![Hierarchy of the traits](media/dyn-type/trait-hierarchy.svg)
+
+Pointers and references `&T` will be represented as the pair `(&(), T::Metadata)` in memory.
+
+`Pointee` is a default bound which *cannot be opt-out* (similar to `std::any::Any`).
+Writing `<T: ?Sized>` would still result in `T: Pointee`,
+except that the members of the traits are not visible until explicitly `use`d.
+As an example, the following should type-check:
+
+```rust ,ignore
+fn g<T: Pointee + ?Sized>();
+fn f<T: ?Sized>() { g::<T>() }
+```
+
+### Core functions
+
+Include the following core functions in `core::ptr::*`, according to [RFC #2580].
+
+```rust ,ignore
+pub fn metadata<T: ?Sized>(ptr: *const T) -> T::Metadata;
+
+impl<T: ?Sized> *const T {
+    pub fn from_raw_parts(data: *const (), meta: T::Metadata) -> Self;
+}
+impl<T: ?Sized> *mut T {
+    pub fn from_raw_parts(data: *mut (), meta: T::Metadata) -> Self;
+}
+impl<T: ?Sized> NonNull<T> {
+    pub fn from_raw_parts(data: NonNull<()>, meta: T::Metadata) -> Self;
+}
+```
+
+Introduce the following new functions in `core::dst::*` as well:
+
+```rust ,ignore
+// Reduction on general custom DSTs are unsafe
+pub unsafe fn reduce_ptr_unchecked<T: Derived + ?Sized>(val: *const T) -> *const Self::Reduced {
+    let ptr = val as *const ();
+    let meta = T::reduce_metadata(val);
+    Self::from_raw_parts(ptr, meta)
+}
+pub unsafe fn reduce_mut_ptr_unchecked<T: Derived + ?Sized>(val: *mut T) -> *mut Self::Reduced { ... }
+pub unsafe fn reduce_non_null_unchecked<T: Derived + ?Sized>(val: NonNull<T>) -> NonNull<Self::Reduced> { ... }
+pub unsafe fn reduce_unchecked<T: Derived + ?Sized>(val: &T) -> &Self::Reduced { ... }
+pub unsafe fn reduce_mut_unchecked<T: Derived + ?Sized>(val: &mut T) -> &mut Self::Reduced { ... }
+
+// Reduction on regular DSTs are safe
+pub fn reduce_ptr<T: Regular + ?Sized>(val: *const T) -> *const Self::Reduced {
+    unsafe { reduce_ptr_unchecked(val) }
+}
+pub fn reduce_mut_ptr<T: Regular + ?Sized>(val: *mut T) -> *mut Self::Reduced { ... }
+pub fn reduce_non_null<T: Regular + ?Sized>(val: NonNull<T>) -> NonNull<Self::Reduced> { ... }
+pub fn reduce<T: Regular + ?Sized>(val: &T) -> &Self::Reduced { ... }
+pub fn reduce_mut<T: Regular + ?Sized>(val: &mut T) -> &mut Self::Reduced { ... }
+```
+
+### Syntax
+
+Introduce a new kind of item.
+
+```rust ,ignore
+ItemKind =
+    ... |
+    CustomDst:{
+        "dyn" "type"
```

Comment:
I think people will get the impression that these are type aliases somehow rather than new nominal types...
You might want to consider `dyn struct`.
Also, consider if you want `dyn(inline) type...` instead and if these could become attributes, e.g.:

```rust
#[repr(C), inline] // name already exists as an attribute so perhaps some other name...
pub dyn struct CStr(u8);
```

(generally if you can please the "no new syntax" people...)

Reply:
The problem (advantage?) of `dyn struct` is it makes it very natural to introduce

```rust
#[repr(C)]
#[dyn(inline)]
pub dyn struct Thin<T> {
    meta: T::Metadata,
    content: T,
}
```

however the type `Thin<T>::Reduced` will point to an anonymous struct 🤔

Reply:
I'm not sure; I'm not nearly an expert on the contents of this RFC ;)

Perhaps?:

```rust
#[repr(C), dyn(inline)]
pub dyn struct CStr = OtherType;
```

Reply:
Lemme just shove all of these into the syntax bikeshed 😂

Reply:
Sounds 🎵 good 🎸